component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java

```diff
@@ -0,0 +1,388 @@
+                attributeExpressionExecutors, attributeStartIndex, noOfDimensions);
+
+        //creating models
+        clusTreeModel = new ClusTreeModel();
+        clusTreeModel.init(noOfDimensions, noOfClusters, maxHeightOfTree, horizon);
+        kMeansModel = new KMeansModel();
+
+        executorService = siddhiAppContext.getExecutorService();
+
+        //setting return attributes
+        List<Attribute> attributeList = new ArrayList<>(1 + noOfDimensions);
+        attributeList.add(new Attribute("euclideanDistanceToClosestCentroid", Attribute.Type.DOUBLE));
+        for (int i = 1; i <= noOfDimensions; i++) {
+            attributeList.add(new Attribute("closestCentroidCoordinate" + i, Attribute.Type.DOUBLE));
+        }
+        return attributeList;
+    }
+
+    @Override
+    protected void process(ComplexEventChunk<StreamEvent> complexEventChunk, Processor processor,
+                           StreamEventCloner streamEventCloner, ComplexEventPopulater complexEventPopulater) {
+        synchronized (this) {
+            while (complexEventChunk.hasNext()) {
+                StreamEvent streamEvent = complexEventChunk.next();
+                noOfEventsReceived++;
+
+                //validating and getting coordinate values
+                for (int i = 0; i < noOfDimensions; i++) {
+                    try {
+                        Number content = (Number) featureVariableExpressionExecutors.get(i).execute(streamEvent);
+                        coordinateValuesOfCurrentDataPoint[i] = content.doubleValue();
+                    } catch (ClassCastException e) {
+                        throw new SiddhiAppRuntimeException("coordinate values should be int/float/double/long " +
+                                "but found " +
+                                featureVariableExpressionExecutors.get(i).execute(streamEvent).getClass());
+                    }
+                }
+
+                //train the ClusTree Model with the datapoint
+                clusTreeModel.trainOnEvent(coordinateValuesOfCurrentDataPoint, null);
+
+                //train the model periodically
+                if (noOfEventsReceived % noOfEventsToRefreshMacroModel == 0) {
+                    LinkedList<DataPoint> dpa = clusTreeModel.getMicroClusteringAsDPArray();
+                    if (noOfEventsToRefreshMacroModel < separateThreadThreshold) {
+                        kMeansModel.refresh(dpa, noOfClusters, maxIterations,
+                                noOfDimensions);
+                    } else {
+                        Trainer trainer = new Trainer(kMeansModel, dpa, noOfClusters, maxIterations, noOfDimensions);
+                        Future f = executorService.submit(trainer);
+                    }
+                }
+
+                //make prediction if the model is trained
+                if (kMeansModel.isTrained()) {
+                    if (logger.isDebugEnabled()) {
+                        logger.debug("Populating output");
```

Comment:
can we add output?

Reply:
sorry i dont understand. please explain

Reply:
basically explain what is output. Always think from the person who sees this log. What can he get to know from this?

Reply:
oh this will populate the event with the prediction once the model is trained. i will add the details into the log